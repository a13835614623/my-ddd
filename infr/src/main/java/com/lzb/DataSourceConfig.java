package com.lzb;

import java.sql.SQLException;

import javax.sql.DataSource;

import org.springframework.boot.jdbc.DataSourceBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class DataSourceConfig {
	@Bean
	public DataSource dataSource() throws SQLException {
		System.out.println("初始化datasource");
		DataSource dataSource = DataSourceBuilder.create().build();
		System.out.println(dataSource.getConnection());
		return dataSource;
	}

}