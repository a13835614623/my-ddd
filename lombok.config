# @AllArgsConstructor 才会生效 replacement @ConstructorProperties({"addressLine1", "addressLine2"})
lombok.anyConstructor.addConstructorProperties=true
lombok.addLombokGeneratedAnnotation = true