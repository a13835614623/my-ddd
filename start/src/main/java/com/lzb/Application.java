package com.lzb;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.EnableAspectJAutoProxy;

@SpringBootApplication
// AopContext.currentProxy() 启用自身调用aop
@EnableAspectJAutoProxy(proxyTargetClass = true, exposeProxy = true)
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}

}
