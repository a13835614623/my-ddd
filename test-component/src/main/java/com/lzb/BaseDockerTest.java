package com.lzb;

import java.time.Duration;
import java.time.Instant;
import java.util.stream.Stream;

import com.redis.testcontainers.RedisContainer;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.TestInfo;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Testcontainers;
import org.testcontainers.utility.DockerImageName;
import org.testcontainers.containers.GenericContainer;

import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.support.TestPropertySourceUtils;

@Slf4j
@Testcontainers(parallel = true)
@ContextConfiguration(initializers = {
        BaseDockerTest.DataSourceInitializer.class,
        BaseDockerTest.RedisInitializer.class
})
public abstract class BaseDockerTest extends BaseTest {


    ///////////////////////////////////////////////////////////////////////////
    // docker相关
    ///////////////////////////////////////////////////////////////////////////

    public static final PostgreSQLContainer<?> database = new PostgreSQLContainer<>("postgres:14.9")
            .withAccessToHost(true)
            .withPrivilegedMode(true)
            .withUsername("postgres")
            .withPassword("123456")
            ;

    public static final RedisContainer redis =
            new RedisContainer(DockerImageName.parse("redis:7.2.0"))
                    .withExposedPorts(6379);

    static class DataSourceInitializer
            implements ApplicationContextInitializer<ConfigurableApplicationContext> {
        @Override
        public void initialize(ConfigurableApplicationContext applicationContext) {
            PostgreSQLContainer<?> database = BaseDockerTest.database;
            TestPropertySourceUtils.addInlinedPropertiesToEnvironment(
                    applicationContext,
                    "spring.datasource.url=" + database.getJdbcUrl(),
                    "spring.datasource.username=" + database.getUsername(),
                    "spring.datasource.password=" + database.getPassword()
            );
        }
    }

    static class RedisInitializer
            implements ApplicationContextInitializer<ConfigurableApplicationContext> {

        @Override
        public void initialize(ConfigurableApplicationContext applicationContext) {
            RedisContainer redis = BaseDockerTest.redis;
            TestPropertySourceUtils.addInlinedPropertiesToEnvironment(
                    applicationContext,
                    "spring.data.redis.host=" + redis.getHost(),
                    "spring.data.redis.port=" + redis.getFirstMappedPort()
            );
        }
    }

    static {

        Instant start = Instant.now();
        Stream.of(database, redis).parallel().forEach(GenericContainer::start);
        log.info("----------------------------------------------------------------------");
        log.info("TestContainers 启动耗时 {} 秒", Duration.between(start, Instant.now()).getSeconds());
        log.info("----------------------------------------------------------------------");

    }

    @AfterAll
    public static void afterAll(TestInfo testInfo) {
    }

    /**
     * 每个test类都会执行一次
     */
    @BeforeAll
    public static void beforeAll(TestInfo testInfo) {
    }

    /**
     * 每个方法执行之前执行一次
     */
    @BeforeEach
    public void beforeEach() {
    }

}