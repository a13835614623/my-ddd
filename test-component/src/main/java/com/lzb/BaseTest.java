package com.lzb;


import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.json.JsonReadFeature;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.mockito.Mockito;

/**
 * <br/>
 * Created on : 2023-03-01 10:31
 * @author lizebin
 */
public abstract class BaseTest extends Mockito implements BaseAssertions {

    private static final ObjectMapper INSTANCE = new JacksonObjectMapper();

    public static class JacksonObjectMapper extends ObjectMapper {

        public JacksonObjectMapper() {
            super();
            //序列化处理
            super.configure(JsonReadFeature.ALLOW_UNESCAPED_CONTROL_CHARS.mappedFeature(), true);
            super.configure(JsonReadFeature.ALLOW_BACKSLASH_ESCAPING_ANY_CHARACTER.mappedFeature(), true);
            super.findAndRegisterModules();
            //失败处理
            super.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false);
            super.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
            //单引号处理
            super.configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true);
            //反序列化时，属性不存在的兼容处理
            super.getDeserializationConfig().withoutFeatures(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
            super.findAndRegisterModules();
            super.configure(SerializationFeature.WRITE_ENUMS_USING_TO_STRING, true);
            super.configure(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY, true);
            super.configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);
            super.configure(SerializationFeature.INDENT_OUTPUT, true);
        }

    }

    protected void assertJSON(Object o) {
        try {
            Approvals.verify(INSTANCE.writerWithDefaultPrettyPrinter().writeValueAsString(o),
                    new Options().forFile().withExtension(".json"));
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

}
