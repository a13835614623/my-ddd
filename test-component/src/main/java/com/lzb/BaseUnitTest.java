package com.lzb;

import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

/**
 * <br/>
 * Created on : 2023-03-01 10:31
 * @author lizebin
 */
@ExtendWith(MockitoExtension.class)
public abstract class BaseUnitTest extends BaseTest {

}
