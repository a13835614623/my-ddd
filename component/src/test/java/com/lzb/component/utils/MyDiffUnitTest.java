package com.lzb.component.utils;

import java.io.Serializable;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

import com.lzb.BaseUnitTest;
import lombok.Data;
import org.javers.core.metamodel.annotation.Id;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

class MyDiffUnitTest extends BaseUnitTest {

    @Test
    @DisplayName("测试空对象")
    void test_diff_empty_object() {

        Person p1 = new Person();
        Person p2 = new Person();

        MyDiff<Person> diff = new MyDiff<>(p1, p2);
        assertThat(diff.hasChanges()).isFalse();
    }

    @Test
    @DisplayName("测试整形对比")
    void test_diff_int() {
        Person p1 = new Person();
        p1.setId(1L);
        p1.age = 10;

        Person p2 = new Person();
        p2.setId(1L);
        p1.age = 11;

        MyDiff<Person> diff = new MyDiff<>(p1, p2);
        assertThat(diff.hasChanges()).isTrue();

        p1.age = 10;
        p2.age = 10;
        diff = new MyDiff<>(p1, p2);
        assertThat(diff.hasChanges()).isFalse();
    }

    @Test
    @DisplayName("测试BigDecimal对比")
    void test_diff_bigdecimal() {
        Person p1 = new Person();
        p1.setId(1L);
        p1.money = BigDecimal.ONE;

        Person p2 = new Person();
        p2.setId(1L);
        p2.money = BigDecimal.TEN;

        MyDiff<Person> diff = new MyDiff<>(p1, p2);
        assertThat(diff.hasChanges()).isTrue();

        p1.money = new BigDecimal("88");
        p2.money = new BigDecimal("88");
        diff = new MyDiff<>(p1, p2);
        assertThat(diff.hasChanges()).isFalse();
    }

    @Test
    @DisplayName("对比列表")
    void should_diff_list() {

        List<Person> oldPersons = new ArrayList<>();
        Person p1 = new Person();
        p1.setId(1L);
        p1.money = BigDecimal.ONE;
        oldPersons.add(p1);
        Person p2 = new Person();
        p2.setId(2L);
        p2.money = BigDecimal.TEN;
        oldPersons.add(p2);

        List<Person> newPersons = new ArrayList<>();
        Person p3 = new Person();
        p3.setId(1L);
        p3.money = new BigDecimal("2");
        newPersons.add(p3);
        Person p4 = new Person();
        p4.setId(2L);
        p4.money = new BigDecimal("11");
        newPersons.add(p4);
        Person person5 = new Person();
        person5.setId(3L);
        person5.setName("person5");
        newPersons.add(person5);

        List<Person> adds = new ArrayList<>();
        List<Person> updates = new ArrayList<>();
        List<Person> removes = new ArrayList<>();
        MyDiff.listChangeFunction(oldPersons, newPersons, Person.class, adds::addAll, updates::addAll, removes::addAll);

        assertThat(adds).hasSize(1);
        assertThat(updates).hasSize(2);
        assertJSON(updates);
        assertThat(removes).hasSize(0);

    }


    @Data
    public static class Person implements Serializable {

        @Id
        private Long id;
        private String name;
        private BigDecimal money;
        private Integer age;

    }

}