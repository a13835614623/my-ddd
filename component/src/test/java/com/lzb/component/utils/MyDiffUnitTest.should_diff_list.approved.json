[ {
  "age" : null,
  "id" : 1,
  "money" : 2,
  "name" : null
}, {
  "age" : null,
  "id" : 2,
  "money" : 11,
  "name" : null
} ]